### PR TITLE
ceph: added rook-ceph-default service account

### PR DESCRIPTION
**Pull request by parth-gr**
When a private docker registry is used and an image pull secret is specified in the chart, the pods with default Service Account fail to pull the image due to authentication issues.
Added rook-ceph-default service account and modify the pods specifications by adding the serviceAccountName.

Closes: https://github.com/rook/rook/issues/6673
Co-authored-by: Tareq Sharafy <tareq.sha@gmail.com>

Signed-off-by: parth-gr <partharora1010@gmail.com>

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

File: cluster/examples/kubernetes/ceph/common.yaml [hunk 1]
```diff
@@ -1137,3 +1137,12 @@ roleRef:
   name: rbd-external-provisioner-runner
   apiGroup: rbac.authorization.k8s.io
 # OLM: END CSI RBD CLUSTER ROLEBINDING
+# OLM: BEGIN SERVICEACCOUNT DEFAULT
+---
+apiVersion: v1
+kind: ServiceAccount
+metadata:
```
Comment (subhamkrai) [inline cluster/examples/kubernetes/ceph/common.yaml:1144]: trailing spaces ...see action error

File: pkg/operator/ceph/cluster/mon/spec_test.go [hunk 2]
```diff
@@ -71,6 +72,8 @@ func testPodSpec(t *testing.T, monID string, pvc bool) {
 	assert.NoError(t, err)
 	assert.NotNil(t, d)
 
+	assert.Equal(t, cephcontroller.DefaultServiceAccount, d.Spec.Template.Spec.ServiceAccountName)
```
Comment (subhamkrai) [inline pkg/operator/ceph/cluster/mon/spec_test.go:75]: remove the extra line above.

File: pkg/operator/ceph/cluster/rbd/spec_test.go [hunk 3]
```diff
@@ -90,6 +92,8 @@ func TestPodSpec(t *testing.T) {
 	assert.Equal(t, 1, len(d.Spec.Template.Spec.Volumes[0].Projected.Sources))
 	assert.Equal(t, 4, len(d.Spec.Template.Spec.Containers[0].VolumeMounts))
 
+	assert.Equal(t, cephcontroller.DefaultServiceAccount, d.Spec.Template.Spec.ServiceAccountName)
```
Comment (subhamkrai) [inline pkg/operator/ceph/cluster/rbd/spec_test.go:95]: same.

File: pkg/operator/ceph/file/mds/spec_test.go [hunk 4]
```diff
@@ -93,6 +95,8 @@ func TestPodSpecs(t *testing.T) {
 	assert.NotNil(t, d)
 	assert.Equal(t, v1.RestartPolicyAlways, d.Spec.Template.Spec.RestartPolicy)
 
+	assert.Equal(t, cephcontroller.DefaultServiceAccount, d.Spec.Template.Spec.ServiceAccountName)
```
Comment (subhamkrai) [inline pkg/operator/ceph/file/mds/spec_test.go:98]: same

File: pkg/operator/ceph/controller/spec.go [hunk 5]
```diff
@@ -46,6 +47,7 @@ const (
 	logCollector                          = "log-collector"
 	DaemonIDLabel                         = "ceph_daemon_id"
 	daemonTypeLabel                       = "ceph_daemon_type"
+	DefaultServiceAccount                 = "rook-ceph-default"
```
Comment (subhamkrai) [inline pkg/operator/ceph/controller/spec.go:50]:
```suggestion
	DefaultServiceAccount                 = "rook-ceph-default"  // DefaultServiceAccount is a service account used for components that do not specify a dedicated service-account.
```

File: pkg/operator/ceph/controller/spec.go [hunk 6]
```diff
@@ -36,6 +36,7 @@ import (
 const (
 	// ConfigInitContainerName is the name which is given to the config initialization container
 	// in all Ceph pods.
+	// DefaultServiceAccount is a  service-account used for components that do not specify a dedicated service-account.
```
Comment (subhamkrai) [inline pkg/operator/ceph/controller/spec.go:39]: remove this

File: pkg/operator/ceph/nfs/spec_test.go [hunk 7]
```diff
@@ -136,4 +138,7 @@ func TestDeploymentSpec(t *testing.T) {
 		},
 	)
 	assert.Equal(t, "my-priority-class", d.Spec.Template.Spec.PriorityClassName)
+
```
Comment (subhamkrai) [inline pkg/operator/ceph/nfs/spec_test.go:141]: same

File: pkg/operator/ceph/nfs/spec_test.go [hunk 8]
```diff
@@ -136,4 +138,7 @@ func TestDeploymentSpec(t *testing.T) {
...
+	assert.Equal(t, cephcontroller.DefaultServiceAccount, d.Spec.Template.Spec.ServiceAccountName)
+
```
Comment (subhamkrai) [inline pkg/operator/ceph/nfs/spec_test.go:143]: same

**Review by subhamkrai — CHANGES_REQUESTED** [review]

File: cluster/examples/kubernetes/ceph/common.yaml (hunk 1 above)
Comment (parth-gr) [reply cluster/examples/kubernetes/ceph/common.yaml]: Done

File: pkg/operator/ceph/nfs/spec_test.go (hunk 8 above)
Comment (parth-gr) [reply pkg/operator/ceph/nfs/spec_test.go:142]: > same

@subhamkrai did the reviewed changes.

File: Documentation/k8s-pre-reqs.md [hunk 9]
```diff
@@ -154,6 +154,14 @@ imagePullSecrets:                # here are the new
 
 After doing this for all service accounts all pods should be able to pull the image from your registry.
 
+### Helm Configuration:
```
Comment (travisn) [inline Documentation/k8s-pre-reqs.md:157]:
```suggestion
### Helm Configuration
```

File: PendingReleaseNotes.md [hunk 10]
```diff
@@ -29,3 +29,4 @@ v1.6...
 * Multiple Ceph mgr daemons are supported for stretch clusters and other clusters where HA of the mgr is more critical
 * Ceph OSD: as of Nautilus 14.2.14 and Octopus 15.2.9 if the OSD scenario is simple (one OSD per disk) we won't use LVM to prepare the disk anymore
 * Disable CSI GRPC metrics by default
+* Added rook-ceph-default Service Account and  configure the default service account pods(mon,mds,etc) specifications with rook-ceph-default Service Account.
```
Comment (travisn) [inline PendingReleaseNotes.md:32]:
```suggestion
* Configure all Rook pods that were using the default service account (mon,mds,etc) to instead use a new `rook-ceph-default` service account.
```

File: pkg/operator/ceph/controller/spec.go (hunk 5 above)
Comment (travisn) [reply pkg/operator/ceph/controller/spec.go:50]: It's preferred to add the comment on the line above instead of at the end

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/controller/spec.go (hunk 5 above)
Comment (subhamkrai) [reply pkg/operator/ceph/controller/spec.go:50]: @parth-gr ^^

**Review by subhamkrai — COMMENTED** [review]
Please squash your commits.Thanks

**Comment by parth-gr** [comment]
> Please squash your commits.Thanks

sure

File: tests/framework/installer/ceph_manifests.go [hunk 11]
```diff
@@ -1028,6 +1028,12 @@ metadata:
 ---
 apiVersion: v1
 kind: ServiceAccount
+metadata:
+  name: rook-ceph-default
+  namespace: ` + namespace + `
+---
+apiVersion: v1
+kind: ServiceAccount
```
Comment (subhamkrai) [inline tests/framework/installer/ceph_manifests.go:1036]: can we do this? directly add a new service account instead of putting adding in middle
```
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: rook-ceph-default
  namespace: ` + namespace + `
---
```

**Review by subhamkrai — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/controller/spec.go (hunk 6 above)
Comment (travisn) [reply pkg/operator/ceph/controller/spec.go:39]: How about moving this comment down to the line immediately before the declaration of `DefaultServiceAccount`?

**Review by travisn — COMMENTED** [review]
LGTM, just one final nit

**Comment by parth-gr** [comment]
> LGTM, just one final nit

done

File: tests/framework/installer/ceph_manifests_v1.5.go [hunk 12]
```diff
@@ -964,6 +964,12 @@ metadata:
 ---
 apiVersion: v1
 kind: ServiceAccount
+metadata:
```
Comment (travisn) [inline tests/framework/installer/ceph_manifests_v1.5.go:967]: Actually, don't set this in the 1.5 test manifests, please revert this file. We want the upgrade to create it as part of the new common.yaml. I believe we'll need a fix to the upgrade test, let's talk about this tomorrow after I merge #7390.

File: pkg/operator/ceph/nfs/spec_test.go (hunk 8 above)
Comment (subhamkrai) [reply pkg/operator/ceph/nfs/spec_test.go:142]: extra line??

File: pkg/operator/ceph/file/mirror/spec_test.go [hunk 13]
```diff
@@ -88,6 +90,8 @@ func TestPodSpec(t *testing.T) {
 	assert.Equal(t, 1, len(d.Spec.Template.Spec.Volumes[0].Projected.Sources))
 	assert.Equal(t, 4, len(d.Spec.Template.Spec.Containers[0].VolumeMounts))
 
+	assert.Equal(t, cephcontroller.DefaultServiceAccount, d.Spec.Template.Spec.ServiceAccountName)
```
Comment (subhamkrai) [inline pkg/operator/ceph/file/mirror/spec_test.go:93]: the extra line above.

**Review by subhamkrai — CHANGES_REQUESTED** [review]
small nits with space only. Thanks

File: tests/framework/installer/ceph_manifests_v1.5.go (hunk 12 above)
Comment (travisn) [reply tests/framework/installer/ceph_manifests_v1.5.go]: With #7390 I added a fix to the upgrade test that should make this scenario already work. After you rebase and fix the merge conflicts the tests should pass.

Comment (parth-gr) [reply tests/framework/installer/ceph_manifests_v1.5.go]: Yes @travisn , you have upgraded the manifests file and manifests_v1.5  with `common.yaml` settings so all the **resources** must be taken from `common.yaml` file, 
I have Rebase my commit but now the tests are failing, I got one point  on #7390 that maybe causing error.

The `common.yaml` in [ceph_manifests.go](https://github.com/rook/rook/blob/master/tests/framework/installer/ceph_manifests.go) file is called by [m.settings.readManifest](https://github.com/rook/rook/blob/3bbc358d321ae3b9dc9ad3dca69eaaabe2d9ced5/tests/framework/installer/ceph_manifests.go#L92) and in its definition [readManifestFromGithubWithClusterNamespace](https://github.com/rook/rook/blob/3bbc358d321ae3b9dc9ad3dca69eaaabe2d9ced5/tests/framework/installer/ceph_settings.go#L82) is called and this function is calling [raw content from github](	https://github.com/rook/rook/blob/3bbc358d321ae3b9dc9ad3dca69eaaabe2d9ced5/tests/framework/installer/ceph_settings.go#L87) so that the issue of calling the raw content from web.

As in the **commit** if their any changes in `common.yaml` file so these are not reflected by this **url** of raw content from github. 
So the integration tests will fail in such cases.

I think linking the file by giving the **path** to the file or  importing the package will help instead of mapping it with the web version.

File: pkg/operator/ceph/file/mirror/spec_test.go (hunk 13 above)
Comment (parth-gr) [reply pkg/operator/ceph/file/mirror/spec_test.go:92]: Already did the nfs changes and updated the mirror.

File: tests/framework/installer/ceph_manifests_v1.5.go (hunk 12 above)
Comment (sp98) [reply tests/framework/installer/ceph_manifests_v1.5.go]: are you seeing the issue being fixed in - https://github.com/rook/rook/pull/7451

Comment (parth-gr) [reply tests/framework/installer/ceph_manifests_v1.5.go]: Hy @sp98 , sorry  but I couldn't relate this to 7451 issue
But what I probaby think [ceph_settings.go](https://github.com/rook/rook/blob/3bbc358d321ae3b9dc9ad3dca69eaaabe2d9ced5/tests/framework/installer/ceph_settings.go) file should be modified and should edit the content of [ line raw content from github](	https://github.com/rook/rook/blob/3bbc358d321ae3b9dc9ad3dca69eaaabe2d9ced5/tests/framework/installer/ceph_settings.go#L87).

File: PendingReleaseNotes.md [hunk 14]
```diff
@@ -29,3 +29,5 @@ v1.6...
 * Multiple Ceph mgr daemons are supported for stretch clusters and other clusters where HA of the mgr is more critical
 * Ceph OSD: as of Nautilus 14.2.14 and Octopus 15.2.9 if the OSD scenario is simple (one OSD per disk) we won't use LVM to prepare the disk anymore
 * Disable CSI GRPC metrics by default
+* Configure all Rook pods that were using the default service account (mon,mds,etc) to instead use a new `rook-ceph-default` service account.
```
Comment (BlaineEXE) [inline PendingReleaseNotes.md:32]: Suggest rewording here to remove "Configure all" which users might interpret as an instruction. 

Suggestion:
Ceph daemon pods using the `default` service account now use a new `rook-ceph-default` service account.

File: pkg/operator/ceph/cluster/cleanup.go [hunk 15]
```diff
@@ -151,6 +152,8 @@ func (c *ClusterController) cleanUpJobTemplateSpec(cluster *cephv1.CephCluster,
 			Volumes:           volumes,
 			RestartPolicy:     v1.RestartPolicyOnFailure,
 			PriorityClassName: cephv1.GetCleanupPriorityClassName(cluster.Spec.PriorityClassNames),
+			// use service account "rook-ceph-default"
```
Comment (BlaineEXE) [inline pkg/operator/ceph/cluster/cleanup.go:155]: Remove unnecessary comment.

File: pkg/operator/ceph/nfs/spec_test.go [hunk 16]
```diff
@@ -20,6 +20,8 @@ import (
 	"testing"
 
 	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	cephcontroller "github.com/rook/rook/pkg/operator/ceph/controller"
+
```
Comment (BlaineEXE) [inline pkg/operator/ceph/nfs/spec_test.go:24]: empty line

**Review by BlaineEXE — COMMENTED** [review]
Mostly nits.

File: pkg/operator/ceph/file/mds/spec_test.go [hunk 17]
```diff
@@ -25,6 +25,8 @@ import (
 	"github.com/rook/rook/pkg/operator/k8sutil"
 
 	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	cephcontroller "github.com/rook/rook/pkg/operator/ceph/controller"
+
```
[the remark this answers is not in the document]
Comment (parth-gr) [reply pkg/operator/ceph/file/mds/spec_test.go:29]: Okay, it was automatically set to this on saving the go file by  (cltr + s), maybe because of go format.

File: cluster/examples/kubernetes/ceph/common.yaml [hunk 18]
```diff
@@ -373,9 +373,17 @@ subjects:
...
 # and bindings accordingly.
 #################################################################################################################
-# Service account for the Ceph OSDs. Must exist and cannot be renamed.
+# OLM: BEGIN SERVICEACCOUNT DEFAULT
+---
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: rook-ceph-default
+  namespace: rook-ceph
```
Comment (subhamkrai) [inline cluster/examples/kubernetes/ceph/common.yaml:382]:
```suggestion
  namespace: rook-ceph # namespace:operator
```
the tests are failing because it was no replaced according to suites namespaces

Comment (parth-gr) [reply cluster/examples/kubernetes/ceph/common.yaml]: Thanks @subhamkrai :+1: , the changes pointed above worked for me :)

**Review by travisn — APPROVED** [review]
LGTM with the green CI. No more changes planned, right?

**Comment by parth-gr** [comment]
> LGTM with the green CI. No more changes planned, right?

Yes @travisn, no more changes planned :+1: 
Just reviewing the code for the final time.

**Review by subhamkrai — APPROVED** [review]
